ts/src/bitget.ts

```diff
@@ -3294,51 +3295,96 @@ export default class bitget extends Exchange {
-        params = this.omit (params, [ 'until', 'till' ]);
+        const defaultLimit = 100; // by default, exchange returns 100 items
+        const msInDay = 1000 * 60 * 60 * 24;
         if (limit !== undefined) {
+            limit = Math.min (limit, maxLimit);
             request['limit'] = limit;
         }
-        if (since !== undefined) {
-            request['startTime'] = since;
+        const until = this.safeInteger2 (params, 'until', 'till');
+        params = this.omit (params, [ 'until', 'till' ]);
+        if (until !== undefined) {
+            request['endTime'] = until;
         }
         if (since !== undefined) {
-            if (limit === undefined) {
-                limit = 100; // exchange default
+            request['startTime'] = since;
+            if (market['spot'] && (until === undefined)) {
+                // for spot we need to send "entTime" too
+                const limitForEnd = (limit !== undefined) ? limit : defaultLimit;
+                const calculatedEnd = this.sum (since, duration * limitForEnd);
+                request['endTime'] = calculatedEnd;
             }
-            const duration = this.parseTimeframe (timeframe) * 1000;
-            request['endTime'] = this.sum (since, duration * (limit + 1)) - 1;  // limit + 1)) - 1 is needed for when since is not the exact timestamp of a candle
-        } else if (until !== undefined) {
-            request['endTime'] = until;
-        } else {
-            request['endTime'] = this.milliseconds ();
         }
         let response = undefined;
-        const thirtyOneDaysAgo = this.milliseconds () - 2678400000;
+        const now = this.milliseconds ();
+        // retrievable periods listed here:
+        // - https://www.bitget.com/api-doc/spot/market/Get-Candle-Data#request-parameters
+        // - https://www.bitget.com/api-doc/contract/market/Get-Candle-Data#description
+        const retrievableDaysMapForSpot = {
+            '1m': 30,
+            '3m': 30,
+            '5m': 30,
+            '10m': 52,
+            '15m': 52,
+            '30m': 52,
+            '1h': 83,
+            '2h': 120,
+            '4h': 240,
+            '6h': 360,
+            '12h': 360,
+            '1d': 360,
+            '3d': 1000,
+            '1w': 1000,
+            '1M': 1000,
+        };
+        const maxRetrievableDaysForNonHistory = this.safeInteger (retrievableDaysMapForSpot, timeframe, 30); // default to safe minimum
+        const endpointTsBoundary = now - maxRetrievableDaysForNonHistory * msInDay;
+        // checks if we need history endpoint
+        let needsHistoryEndpoint = false;
+        if (since !== undefined && since < endpointTsBoundary) {
+            needsHistoryEndpoint = true;
+        } else if (until !== undefined && ((limit === undefined && until < endpointTsBoundary) || (limit !== undefined && until - limit * duration < endpointTsBoundary))) {
+            needsHistoryEndpoint = true;
```

Comment:
this is not simple at all and very hard to understand

Reply:
i will dissect them into variables and lyk then

Reply:
i've simplified and more readable now

Reply:
that was the reason when I mentioned in discord, that i wanted to avoid `if limit is set by user` checks by defaulting it, but eventually i avoided, so it's flexible now and doesn't enforce any defaults.